added snake case feature and tests

cathulhu opened the pull request:
This is pull request for issue #94.

cathulhu committed:
added snake case feature, tests, and updated pre-commit version

shreyb commented on .gitignore:
```diff
@@ -7,3 +7,4 @@ __pycache__
 swagger.json
 config/swagger.json
 remove
+.vscode
```
Good call.

shreyb commented on ferry_cli/__main__.py:
```diff
@@ -545,6 +545,27 @@ def handle_no_args(_config_path: Optional[pathlib.Path]) -> bool:
+        endpoint_lowercase = arguments[1].lower()
+
+        # check for underscore seperated endpoint and convert it
+        if "_" in arguments[1]:
```
I believe you don't need this `if` check here.  `str.replace()` should just leave the string untouched if it doesn't have an `_` in it:

```python
>>> "blah_blah".replace("_","")
'blahblah'
>>> "blahBlahBlah".replace("_","")
'blahBlahBlah'
```

shreyb commented on ferry_cli/__main__.py:
```diff
@@ -545,6 +545,27 @@ def handle_no_args(_config_path: Optional[pathlib.Path]) -> bool:
+        if "_" in arguments[1]:
+            endpoint_lowercase = endpoint_lowercase.replace("_", "")
+
+        for entry in endpoints:
```
Given that `endpoints` is a dict, it'd be nice if we can look up `endpoint_lowercase` in `endpoints` rather than iterating through until we find the matching entry (or not).  But for now, this certainly works.

shreyb replied:
The more I think about it, especially since this isn't a time-pressing issue, the more I feel we should really try to get this right.  

So to summarize, this code would be a lot more efficient if instead of iterating through a set of dict keys, normalizing each one, and comparing it to a normalized candidate key; we mutate the candidate key to the format we expect, and then just do a dict lookup.

That is to say, if the user gives something like `-e get_User_info`, we KNOW they mean `getUserInfo`, and if they don't, we should treat their input as invalid.  We could mutate the `get_User_info` to `getUserInfo`, and just do a membership check (i.e. `if mutated_key in endpoints...`).  That's a lot quicker from a performance standpoint, and I suspect it'll tighten up the code too.

shreyb commented on tests/test_main.py:
```diff
@@ -302,3 +303,20 @@ def test_handle_no_args_configfile_does_not_exist(
+def test_snakecase_and_underscore_conversion():
+    test_endpoints = ["getUserInfo"]
+    correct_args = ["-e", "getUserInfo", "--username=johndoe"]
+    test_args_case_underscore = ["-e", "_Get_USeriNFo", "--username=johndoe"]
```
Do we want to support the leading underscore?  I know the code as you have it will pass this test, but we should probably say in the documentation that we only support snake case where it makes sense (i.e. getUserInfo, get_user_info, and get_User_Info are all supported, but _get_User_Info_ are not supported, even if they actually will work).

shreyb reviewed (changes requested):
This looks good.  I just had the one request for a change, and one general comment.  The latter can be worked on in another issue.

shreyb commented on ferry_cli/__main__.py:
```diff
@@ -545,6 +545,27 @@ def handle_no_args(_config_path: Optional[pathlib.Path]) -> bool:
+) -> List[str]:
+
+    # check to see if the arguments supplied are for an endpoint. IE a "-e" was supplied
+    if arguments[0] == "-e" or arguments[0] == "-E":
```
I was chatting with Kevin, and this PR came up.  He pointed out that this logic works if the user gives us exactly `ferry-cli -e ENDPOINT`, but if they don't, it becomes fragile.  To demonstrate this, go ahead and try running a malformed command like:
```
ferry-cli --foo -e get_user_info --username=nosuchuser
```

In that case, `arguments[0]` is actually equal to `--foo`, and so the intended transformation of `get_user_info` to `getUserInfo` won't happen, and the error you get is actually:

```
An error occurred while using the FERRY CLI: Error: 'get_user_info' is not a valid endpoint. Run 'ferry -l' for a full list of available endpoints.
```
which is the wrong error to print here.

There are a couple of possible ways to fix this that I can think of off the top of my head:
1) Use a function-scoped `argparse.ArgumentParser` JUST in this function, that only has the argument `-e` defined.  Then use `ArgumentParser.parse_known_args()` on a copy of `arguments` to make sure you save all the arguments and can return an argument list with just the `-e` value replaced.
2) Iterate through the `arguments` list, and once you find your `-e` or `-E` (or `--endpoint`), do the value transformation.

retzkek replied:
I'm going to throw another wrench into the works:

> and once you find your -e or -E (or --endpoint)

What if that changes? In five years someone comes along and says "-e/-E/--endpoint" is a terrible name, it should be -p/--point-final 🇫🇷." How does someone changing the main parser (on line 109) know to change this too (on line 548)?

shreyb replied:
I mean, at some point we can't solve for every possible future situation - we have to make some assumptions.

But to safeguard against this, Lydia, you could add a test to make sure that calls both `FerryCLI.get_arg_parser` and this function on the same set of args, and make sure both return expected values.  Maybe there's a more elegant way than that, but something that ensures that what you're checking for here is a valid `FerryCLI` argument.

retzkek replied:
The point I hoped to get across (perhaps too subtly) was that you should first look for a way to solve this that **doesn't even need to know or care what the actual flags are**. Even better if you don't need to do some second-level parsing of arguments at all. A couple ideas:

- Use a custom `action` class for the argument. https://docs.python.org/3/library/argparse.html#action-classes
- Maybe you can add the functionality to the `FerryParser`.

shreyb replied:
Too subtle! 

But in all seriousness, I _do_ like the idea of having a custom `Action` for the `--endpoint` argument that just does this transformation on the argument automatically (or leaves it alone).  @cathulhu - have you worked with custom `Actions`?  

If not, we have a couple of examples in the `FerryCLI` class (`get_endpoint_params_action`), but to see a more direct, less-wrapped up example, here's one from jobsub where we store the values of the `-G` flag there in the environment:  https://github.com/fermitools/jobsub_lite/blob/master/lib/get_parser.py#L42.

LTrestka committed:
Handled error correction for endpoint argument.
- Converts snake case to lower camel case.
- Handles edge case where proper endpoint is passed, but uses incorrect capitalization.

LTrestka committed:
Runs same check for endpoint params argument

shreyb commented on ferry_cli/__main__.py:
```diff
@@ -548,21 +548,20 @@ def handle_no_args(_config_path: Optional[pathlib.Path]) -> bool:
+    for i in range(len(arguments)):
+        if (arguments[i].lower() == "-e" or arguments[i].lower() == "--endpoint") and len(arguments) > i + 1:
+            # convert snake_case_endpoint arg to lowerCamelCase
+            arguments[i+1] = "".join([part.capitalize() for part in arguments[i+1].split("_")])
```
Please put a comment here saying what this line is doing.  (I know it's splitting on "_", and then capitalizing each part, but let's make it clear to the future reader)

shreyb commented on ferry_cli/__main__.py:
```diff
@@ -548,21 +548,20 @@ def handle_no_args(_config_path: Optional[pathlib.Path]) -> bool:
+            # convert snake_case_endpoint arg to lowerCamelCase
+            arguments[i+1] = "".join([part.capitalize() for part in arguments[i+1].split("_")])
+            if len(arguments[i+1]) > 0:
+                arguments[i+1] = arguments[i+1][:1].lower() + arguments[i+1][1:]
```
Same here - please explain that we're lower-casing the first letter.

shreyb commented on ferry_cli/__main__.py:
```diff
@@ -548,21 +548,20 @@ def handle_no_args(_config_path: Optional[pathlib.Path]) -> bool:
-    # otherwise if not endpoint just return back the same set of args we got to be handled by rest of the program
+    for i in range(len(arguments)):
+        if (arguments[i].lower() == "-e" or arguments[i].lower() == "--endpoint") and len(arguments) > i + 1:
+            # convert snake_case_endpoint arg to lowerCamelCase
```
I think everything from this line through 558 should really be its own function that returns the transformed argument, to clean up the code space.  Then this for-loop would read:

```python
for i in range(len(arguments)):
  if (arguments[i].......):
    arguments[i+1] = snake_case_to_lowerCamelCase(arguments[i+1]) # I just made this name up. Please use something better
   # handle case where user supplies endpoint with improper capitalization
   if arguments[i+1] not in endpoints:
      # stuff
return arguments
```

shreyb replied:
On second thought, I think we shouldn't even assign `arguments[i+1]` here, in case it fails the `if .... not in endpoints` check.  Maybe assign it to some temporary variable (`candidate_endpoint`) or something like that, and then assign that to `arguments[i+1]` if it passes all the checks?

shreyb commented on ferry_cli/__main__.py:
```diff
@@ -548,21 +548,20 @@ def handle_no_args(_config_path: Optional[pathlib.Path]) -> bool:
+            if len(arguments[i+1]) > 0:
+                arguments[i+1] = arguments[i+1][:1].lower() + arguments[i+1][1:]
+            
+            # handle case where user supplies endpoint with improper capitalization
```
Rather than check for improper capitalization here, I think we should transform the user input above (like on line 555).  That could read:

```python
arguments[i+1] = "".join([part.lower().capitalize() for part in arguments[i+1].split("_")])
```

Check my work on that, but I _think_ that would make every part lowercased, then capitalize it.  If so, we wouldn't need this check at all.

I also think that would make it clearer if we reimplement Lydia's logic into this part; so assuming we did what I suggested earlier in this comment, this portion would look like:

```python
if candidate_argument not in endpoints:   # at this point, arguments[i+1] should be the correct capitalization
  return arguments    # if you don't change arguments until after this check passes, this works
arguments[i+1] = candidate_argument # do the assignment because the candidate_argument passes the test
return arguments
```

shreyb commented on ferry_cli/__main__.py:
```diff
@@ -548,21 +548,20 @@ def handle_no_args(_config_path: Optional[pathlib.Path]) -> bool:
-                return arguments
-
-    # otherwise if not endpoint just return back the same set of args we got to be handled by rest of the program
+    for i in range(len(arguments)):
```
Generally, the more pythonic practice these days is to write this line:

```python
for i, arg in enumerate(arguments)
```

Then below, all the `arguments[i]` can become `arg`.  The `arguments[i+1]` would stay as is.

shreyb commented on ferry_cli/__main__.py:
```diff
@@ -548,21 +548,20 @@ def handle_no_args(_config_path: Optional[pathlib.Path]) -> bool:
-
-    # otherwise if not endpoint just return back the same set of args we got to be handled by rest of the program
+    for i in range(len(arguments)):
+        if (arguments[i].lower() == "-e" or arguments[i].lower() == "--endpoint") and len(arguments) > i + 1:
```
Two comments on this line:

1. The first two checks can be combined to `if arguments[i].lower() in ("-e", "--endpoint")`. Or use `arg` in place of `arguments[i]`, if you take my suggestion above.

2. I love that last check.  Could you please make it clearer by having it be `and i < len(arguments) - 1`? I think that communicates more clearly that we're making sure that we're not on the last argument in this check.

shreyb replied:
Oops.  Ignore (1) here.  Looks like you already did that with a set literal.  Cool.

shreyb commented on ferry_cli/__main__.py:
```diff
@@ -548,21 +548,20 @@ def handle_no_args(_config_path: Optional[pathlib.Path]) -> bool:
+                    if arguments[i+1].lower() == endpoint.lower():
+                        arguments[i+1] = endpoint
+            break
     return arguments
```
I left a lot of comments above, so I thought I'd write a summary comment here.  If you combine all of my suggestions above, you'd end up with a function that looked like this:

```python
def handle_arg_capitalization(endpoints: Dict[str, Any], arguments: List[str]) -> List[str]:
  # check to see if the arguments supplied are for an endpoint. IE a "-e" was supplied
  for i, arg in enumerate(arguments):
    if arg.lower() in ("-e", "--endpoint") and i < len(arguments) - 1:
      candidate_endpoint = snake_case_to_lowerCamelCase(arguments[i+1])

      if candidate_endpoint not in endpoints:
        return arguments
     
     arguments[i+1] = candidate_endpoint
     return arguments
```

And then `snake_case_to_lowerCamelCase` would do all the transformation (and correcting of capitalization).

shreyb commented:
We revisited this in the 7/1/25 meeting, and decided that this was still worth completing.

LTrestka committed:
Merge branch 'master' into lydia_snake_case

LTrestka committed:
updated _sanitize_base_url to handle edge cases where --server path doesnt end with trailing slash.

LTrestka committed:
path needed to strip("/")

LTrestka committed:
Added failed call handling

LTrestka committed:
Keeps leading underscores. Updated Readme with that bit

LTrestka committed:
fixed unit test

shreyb commented on ferry_cli/__main__.py:
```diff
@@ -585,6 +602,36 @@ def handle_no_args(_config_path: Optional[pathlib.Path]) -> bool:
+            ep = leading_underscore + camel
+
+            # Match endpoint case-insensitively and replace original argument if found
+            matched_ep = next((e for e in endpoints if e.lower() == ep.lower()), None)
```
Do we need to match case-insentively? `ep` should already match an `endpoints` key exactly (or not), so why can't this just be a dict lookup rather than an iteration/filter?

i.e. - it'd be much cleaner IMO if this were
```python
if ep in endpoints:
  arguments[idx + 1] = ep
```

shreyb replied:
Also, have you thought about Kevin's suggestion (https://github.com/fermitools/Ferry-CLI/pull/97#discussion_r1929179848)  of handling this endpoint case as a custom action assigned to the `-e/--endpoint` flag?

The other possibility is that at line 360 (https://github.com/fermitools/Ferry-CLI/blob/e1b4f40f575a071f77a0b1466ed70e91c14b7743/ferry_cli/__main__.py#L360), you could even call this new function `handle_arg_capitalization`, but since that's within the `if args.endpoint`, you could just directly pass the value to this function and not even worry about iterating through all the arguments.  Something like, on line 360:

```python
if args.endpoint:
  ep = handle_arg_capitalization(self.endpoints, args.endpoint)
  self.safeguards.verify(ep)
  # ...
```

And then the function `handle_arg_capitalization` could even just be something along the lines of:

```python
def handle_arg_capitalization( endpoints: Dict[str, Any], arg: str) -> str:
  # You can skip all the enumeration
  # Extract and preserve a single leading underscore, if any
            leading_underscore = "_" if raw_arg.startswith("_") else ""
  # ...
```

What do you think about either of those ideas?

shreyb reviewed (changes requested):
This is certainly fine as is, but I think we could make a couple of more tweaks to really tighten up the code.

LTrestka committed:
- addressed comments
  - renamed function to "normalize_endpoint"
  - removes args enumeration
  - called in endpoint/endpoint_params handler before safeguard.

shreyb commented on tests/test_main.py:
```diff
@@ -308,6 +309,23 @@ def test_handle_no_args_configfile_does_not_exist(
+    test_endpoints = ["getUserInfo"]
+    correct_args = ["-e", "getUserInfo", "--username=johndoe"]
+    test_args_case_underscore = ["-e", "Get_USeriNFo", "--username=johndoe"]
+    result = handle_arg_capitalization(test_endpoints, test_args_case_underscore)
```
These instances need to be changed to the new name, `normalize_endpoint`.

LTrestka committed:
updated unit tests for normalized endpoints

shreyb commented:
LGTM
